Sharding Scaling dataSource connection is not released when execute check scaling xxxx command

5.0.0-RC ShardingSphere-Proxy

dataSource connection is not released

log：
[ERROR] 2021-11-02 19:00:02.427 [_finished_check_Worker-1] o.a.s.s.core.job.FinishedCheckJob - scaling job 662316049231773697 finish check failed!
org.apache.shardingsphere.scaling.core.common.exception.DataCheckFailException: get table columns failed
	at org.apache.shardingsphere.scaling.core.job.check.consistency.DataConsistencyCheckerImpl.getTablesColumnNamesMap(DataConsistencyCheckerImpl.java:163)
	at org.apache.shardingsphere.scaling.core.job.check.consistency.DataConsistencyCheckerImpl.dataCheck(DataConsistencyCheckerImpl.java:115)
	at org.apache.shardingsphere.scaling.core.api.impl.ScalingAPIImpl.dataConsistencyCheck0(ScalingAPIImpl.java:190)
	at org.apache.shardingsphere.scaling.core.api.impl.ScalingAPIImpl.dataConsistencyCheck(ScalingAPIImpl.java:174)
	at org.apache.shardingsphere.scaling.core.job.FinishedCheckJob.dataConsistencyCheck(FinishedCheckJob.java:73)
	at org.apache.shardingsphere.scaling.core.job.FinishedCheckJob.execute(FinishedCheckJob.java:59)
	at org.apache.shardingsphere.elasticjob.simple.executor.SimpleJobExecutor.process(SimpleJobExecutor.java:33)
	at org.apache.shardingsphere.elasticjob.simple.executor.SimpleJobExecutor.process(SimpleJobExecutor.java:29)
	at org.apache.shardingsphere.elasticjob.executor.ElasticJobExecutor.process(ElasticJobExecutor.java:177)
	at org.apache.shardingsphere.elasticjob.executor.ElasticJobExecutor.process(ElasticJobExecutor.java:147)
	at org.apache.shardingsphere.elasticjob.executor.ElasticJobExecutor.execute(ElasticJobExecutor.java:130)
	at org.apache.shardingsphere.elasticjob.executor.ElasticJobExecutor.execute(ElasticJobExecutor.java:106)
	at org.apache.shardingsphere.elasticjob.lite.internal.schedule.LiteJob.execute(LiteJob.java:35)
	at org.quartz.core.JobRunShell.run(JobRunShell.java:202)
	at org.quartz.simpl.SimpleThreadPool$WorkerThread.run(SimpleThreadPool.java:573)
Caused by: java.sql.SQLSyntaxErrorException: SELECT command denied to user 'test'@'' for table 'checksums'
	at com.mysql.cj.jdbc.exceptions.SQLError.createSQLException(SQLError.java:120)
	at com.mysql.cj.jdbc.exceptions.SQLExceptionsMapping.translateException(SQLExceptionsMapping.java:122)
	at com.mysql.cj.jdbc.StatementImpl.executeQuery(StatementImpl.java:1202)
	at com.mysql.cj.jdbc.DatabaseMetaData$2.forEach(DatabaseMetaData.java:2159)
	at com.mysql.cj.jdbc.DatabaseMetaData$2.forEach(DatabaseMetaData.java:2088)
	at com.mysql.cj.jdbc.IterateBlock.doForAll(IterateBlock.java:56)
	at com.mysql.cj.jdbc.DatabaseMetaData.getColumns(DatabaseMetaData.java:2264)
	at com.zaxxer.hikari.pool.ProxyDatabaseMetaData.getColumns(ProxyDatabaseMetaData.java:90)
	at com.zaxxer.hikari.pool.HikariProxyDatabaseMetaData.getColumns(HikariProxyDatabaseMetaData.java)
	at org.apache.shardingsphere.driver.jdbc.core.datasource.metadata.ShardingSphereDatabaseMetaData.getColumns(ShardingSphereDatabaseMetaData.java:121)
	at org.apache.shardingsphere.scaling.core.job.check.consistency.DataConsistencyCheckerImpl.getTablesColumnNamesMap(DataConsistencyCheckerImpl.java:156)
	... 14 common frames omitted

Hi @ccw55ccw ,  it seems no `SELECT` permission for `test` account from error log:
> Caused by: java.sql.SQLSyntaxErrorException: SELECT command denied to user 'test'@'' for table 'checksums'

Could you check MySQL `test` account's permissions, e.g. execute following SQL:
```
show grants for 'test'@'';
```

> Hi @ccw55ccw , it seems no `SELECT` permission for `test` account from error log:
> 
> > Caused by: java.sql.SQLSyntaxErrorException: SELECT command denied to user 'test'@'' for table 'checksums'
> 
> Could you check MySQL `test` account's permissions, e.g. execute following SQL:
> 
> ```
> show grants for 'test';
> ```

mysql> show grants for 'test';
+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Grants for test@%                                                                                                                                                                                                                                                                                                                                                |
+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| GRANT INSERT, UPDATE, DELETE, CREATE, DROP, RELOAD, SHUTDOWN, PROCESS, FILE, REFERENCES, INDEX, ALTER, SHOW DATABASES, SUPER, CREATE TEMPORARY TABLES, LOCK TABLES, EXECUTE, REPLICATION SLAVE, REPLICATION CLIENT, CREATE VIEW, SHOW VIEW, CREATE ROUTINE, ALTER ROUTINE, CREATE USER, EVENT, TRIGGER, CREATE TABLESPACE ON *.* TO 'test'@'%' WITH GRANT OPTION |

--- a/pom.xml
+++ b/pom.xml
@@ -72,7 +72,7 @@
         <commons-io.version>2.8.0</commons-io.version>
         
         <javax.transaction.version>1.1</javax.transaction.version>
-        <atomikos.version>4.0.6</atomikos.version>
+        <atomikos.version>5.0.8</atomikos.version>
         <seata.version>1.4.2</seata.version>
         <narayana.version>5.9.1.Final</narayana.version>
         <jboss-transaction-spi.version>7.6.0.Final</jboss-transaction-spi.version>

--- a/shardingsphere-distribution/shardingsphere-proxy-distribution/src/main/release-docs/LICENSE
+++ b/shardingsphere-distribution/shardingsphere-proxy-distribution/src/main/release-docs/LICENSE
@@ -298,11 +298,11 @@ The following components are provided under the Apache License. See project link
 The text of each license is also included at licenses/LICENSE-[project].txt.
 
     groovy 2.4.19-indy: https://github.com/apache/groovy, Apache 2.0
-    atomikos-util 4.0.6: https://www.atomikos.com, Apache 2.0
-    transactions 4.0.6: https://www.atomikos.com, Apache 2.0
-    transactions-api 4.0.6: https://www.atomikos.com, Apache 2.0
-    transactions-jdbc 4.0.6: https://www.atomikos.com, Apache 2.0
-    transactions-jta 4.0.6: https://www.atomikos.com, Apache 2.0
+    atomikos-util 5.0.8: https://www.atomikos.com, Apache 2.0
+    transactions 5.0.8: https://www.atomikos.com, Apache 2.0
+    transactions-api 5.0.8: https://www.atomikos.com, Apache 2.0
+    transactions-jdbc 5.0.8: https://www.atomikos.com, Apache 2.0
+    transactions-jta 5.0.8: https://www.atomikos.com, Apache 2.0
     
 ========================================================================
 BSD licenses

--- a/shardingsphere-kernel/shardingsphere-transaction/shardingsphere-transaction-type/shardingsphere-transaction-xa/shardingsphere-transaction-xa-core/pom.xml
+++ b/shardingsphere-kernel/shardingsphere-transaction/shardingsphere-transaction-type/shardingsphere-transaction-xa/shardingsphere-transaction-xa-core/pom.xml
@@ -37,18 +37,6 @@
             <artifactId>shardingsphere-transaction-xa-atomikos</artifactId>
             <version>${project.version}</version>
         </dependency>
-        <dependency>
-            <groupId>com.atomikos</groupId>
-            <artifactId>transactions</artifactId>
-        </dependency>
-        <dependency>
-            <groupId>com.atomikos</groupId>
-            <artifactId>transactions-jta</artifactId>
-        </dependency>
-        <dependency>
-            <groupId>com.atomikos</groupId>
-            <artifactId>transactions-jdbc</artifactId>
-        </dependency>
         
         <dependency>
             <groupId>com.zaxxer</groupId>

--- a/shardingsphere-proxy/shardingsphere-proxy-bootstrap/src/main/resources/logback.xml
+++ b/shardingsphere-proxy/shardingsphere-proxy-bootstrap/src/main/resources/logback.xml
@@ -26,17 +26,11 @@
         <appender-ref ref="console" />
     </logger>
     
-    <logger name="com.zaxxer.hikari" level="error">
-        <appender-ref ref="console" />
-    </logger>
+    <logger name="com.zaxxer.hikari" level="error" />
     
-    <logger name="com.atomikos" level="error">
-        <appender-ref ref="console" />
-    </logger>
+    <logger name="com.atomikos" level="error" />
     
-    <logger name="io.netty" level="error">
-        <appender-ref ref="console" />
-    </logger>
+    <logger name="io.netty" level="error" />
     
     <root>
         <level value="info" />

--- a/shardingsphere-scaling/shardingsphere-scaling-core/src/main/java/org/apache/shardingsphere/scaling/core/common/datasource/DataSourceWrapper.java
+++ b/shardingsphere-scaling/shardingsphere-scaling-core/src/main/java/org/apache/shardingsphere/scaling/core/common/datasource/DataSourceWrapper.java
@@ -18,6 +18,7 @@
 package org.apache.shardingsphere.scaling.core.common.datasource;
 
 import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
 
 import javax.sql.DataSource;
 import java.io.PrintWriter;
@@ -30,6 +31,7 @@ import java.util.logging.Logger;
  * Data source wrapper is for abstract standard jdbc and sharding jdbc.
  */
 @RequiredArgsConstructor
+@Slf4j
 public final class DataSourceWrapper implements DataSource, AutoCloseable {
     
     private final DataSource dataSource;
@@ -94,6 +96,8 @@ public final class DataSourceWrapper implements DataSource, AutoCloseable {
                 // CHECKSTYLE:ON
                 throw new SQLException("data source close failed.", ex);
             }
+        } else {
+            log.error("dataSource is not closed, it might cause connection leak, dataSource={}", dataSource, new RuntimeException());
         }
     }
 }

